[API] Add bitcast API

zzzDavid opened the pull request:
This pull request is linked to issue #310.

## Description

Adding support for bitcast in `python/heterocl/compute_api.py`: `heterocl.bitcast()`
corresponding test: `tests/test_api_bitcast.py`

seanlatias reviewed (changes requested):
Could you rename `test_bitcast.py` to `test_api_bitcast.py`?

seanlatias commented on python/heterocl/compute_api.py:
```diff
@@ -10,9 +10,12 @@
 from .schedule import Stage
 from .debug import APIError
-from .dsl import if_, for_
+from .dsl import if_, for_, def_
```
We don't need `def_`.

zzzDavid replied:
Removed

seanlatias commented on python/heterocl/compute_api.py:
```diff
@@ -10,9 +10,12 @@
+from .dsl import if_, for_, def_
 from .mutator import Mutator
 from .module import Module
+from .types import Type
+from . import util
+from itertools import product
```
Double-check whether these are indeed used or not.

zzzDavid replied:
They are used indeed in the bitcast API. `util` and `Type` are used to check the input destination datatype and bitwidth, `product` is used to iterate through the input tensor

seanlatias replied:
Please see [line 8 and 10](https://github.com/cornell-zhang/heterocl/blob/bbfdae4abbbe9268c65cdcad7b8d312412cf1248/python/heterocl/compute_api.py#L8). `util` and `types` are already imported. For `itertools` move it to the top along with `numbers` and `numpy`.

seanlatias commented on python/heterocl/compute_api.py:
```diff
@@ -921,5 +924,67 @@ def reduce_body():
+    dst_dtype_str = dtype_to_str(dst_dtype)
+
+    def _iter_tensor(_tensor, tensor, buffer_var):
+        for indices in list(product(*[range(x) for x in tensor.shape])):
+            index, _, _ = get_index(shape, indices, 0)
+            load = _make.Load(tensor.dtype, tensor.buf.data, index)
+            expr = _make.Call(dst_dtype_str, "bitcast", [tensor[indices]], _expr.Call.PureIntrinsic, None, 0)
+            stage.emit(
+                _make.Store(
+                    buffer_var,
+                    _make.Cast(stage._dtype, expr),
+                    index
+                )
+            )
+
```
You should create a `for` loop instead.  [Here](https://github.com/cornell-zhang/heterocl/blob/bbfdae4abbbe9268c65cdcad7b8d312412cf1248/python/heterocl/compute_api.py#L391) is an example of using `make_for`. Otherwise, you are basically unrolling the loop by default.

zzzDavid commented:
@seanlatias Hi, I rebased and re-implemented bitcast API with compute, also added support for bitcasting expression. Please review the changes, thanks!

seanlatias commented on tests/test_api_bitcast.py:
```diff
@@ -0,0 +1,84 @@
+    s = hcl.create_schedule([A], algorithm)
+    f = hcl.build(s)
+
+    print(hcl.lower(s))
```
Remove all `print(hcl.lower(s))` in tests.

zzzDavid replied:
Done

seanlatias commented on python/heterocl/compute_api.py:
```diff
@@ -921,5 +921,57 @@ def reduce_body():
+
+    Parameters
+    ----------
+    tensor: Tensor or Expr
+        The input tensor or expression of the source data type
+    dst_dtype: Type
+        The destination data type. For example, hcl.UInt(32)
+    Name: str
+        The name of the returned tensor
```
Add an empty line between the arguments.

zzzDavid replied:
Done

seanlatias commented on python/heterocl/compute_api.py:
```diff
@@ -921,5 +921,57 @@ def reduce_body():
+        The input tensor or expression of the source data type
+    dst_dtype: Type
+        The destination data type. For example, hcl.UInt(32)
+    Name: str
```
```suggestion
    name: str, optional
```

zzzDavid replied:
Done

seanlatias commented on python/heterocl/compute_api.py:
```diff
@@ -921,5 +921,57 @@ def reduce_body():
+
+    # set up name, shape, and fcompute
+    dst_dtype_str = dtype_to_str(dst_dtype)
+    if isinstance(tensor, Tensor):
+        name = tensor.name + '_' + dst_dtype_str if name is None else name
+        shape = tensor.shape
+        fcompute = lambda *args : _make.Call(dst_dtype_str, "bitcast", [tensor[args]], _expr.Call.PureIntrinsic, None, 0) 
+    else: # expression
+        name = get_name("compute", name)
+        shape = (1,)
+        fcompute = lambda *args : _make.Call(dst_dtype_str, "bitcast", [tensor], _expr.Call.PureIntrinsic, None, 0)
```
Extract `fcompute` out from the if/else statement.

seanlatias commented on python/heterocl/compute_api.py:
```diff
@@ -921,5 +921,57 @@ def reduce_body():
+    src_bitwidth = get_type(tensor.dtype)[1]
+    dst_bitwidth = dst_dtype.bits 
+    if src_bitwidth != dst_bitwidth:
+        if isinstance(tensor, Tensor):
+            raise APIError(f"Bitcasting source Tensor of bitwidth {src_bitwidth} to destination bitwidth {dst_bitwidth}.") 
+        else:
+            raise APIError(f"Bitcasting source expression of bitwidth {src_bitwidth} to destination bitwidth {dst_bitwidth}.")
```
The error message does not tell the right thing. We should at least say the bitwidths are mismatched. Then, we can provide additional info.

zzzDavid replied:
Rewrote the error message

seanlatias commented:
Also, for the expression one, actually, I mean something like this.

```python
a = # some hcl expression
b = hcl.bitcast(a, dst_type) # b is still an expression, instead of a tensor
```

In other words, if the input is an expression, return an expression.

seanlatias commented:
@zzzDavid if you have time can you fix this? It shouldn't take too much time. Thanks.

zzzDavid commented:
> @zzzDavid if you have time can you fix this? It shouldn't take too much time. Thanks.

Should have fixed that sooner, thanks for reminding me. Please review my lastest commit :)

seanlatias reviewed (approved):
LGTM
